### PR TITLE
Must declare variables before assigning to them

### PR DESCRIPTION
**Pull request by joydeep-b**

### CODE REVIEW

File: test/unit-tests.test.ts [hunk 1]
```diff
@@ -532,7 +539,7 @@ test('Can set fields of this in a constructor',  (done) => {
   expect.assertions(2);
   const runner = compileOK(`
     class C { constructor() { this.x = 5; } }
-    r = (new C()).x`);
+    let r = (new C()).x`);
```
Comment (joydeep-b) [inline test/unit-tests.test.ts]: The new static check caught this bug in our test code. :)

**Comment by baxtersa** [comment]
This looks good! with this you could still refer to undeclared variables outside of assignment lvals (it'll throw `ReferenceError` I think if if it's not a global), but you can't assign to them and implicitly declare globals, which I think is all you need.

**Comment by arjunguha** [comment]
Oh, good point. I was worrying about the non-problem that Sam describes. This is much simpler than I realized. OK. Let's not merge this until the testing commit is merged.

**Comment by joydeep-b** [comment]
Great, thanks, Sam. 
We'll merge after project 1.

**Pull request merged**